lib/plugins/aws/utils/request.js

```diff
@@ -0,0 +1,208 @@
+          });
+      };
+      return doCall(0);
+    });
+
+  // Support S3 Transfer Acceleration
+  if (
+    canUseS3TransferAcceleration(service, method) &&
+    serviceOptions.isS3TransferAccelerationEnabled
+  ) {
+    enableS3TransferAcceleration(serviceOptions.credentials);
+  }
+
+  const request = requestQueue.add(() =>
+    persistentRequest(() => {
+      const Service = _.get(sdk, service);
+      serviceOptions.credentials.region = serviceOptions.region;
+      const awsService = new Service(serviceOptions.credentials);
+      const req = awsService[method](params);
+
+      // TODO: Add listeners, put Debug statements here...
+      // req.on('send', function (r) {console.log(r)});
+
+      const promise = req.promise
+        ? req.promise()
+        : BbPromise.fromCallback((cb) => {
+            req.send(cb);
+          });
+      return promise.catch((err) => {
+        let message = err.message != null ? err.message : String(err.code);
+        if (message.startsWith('Missing credentials in config')) {
+          // Credentials error
+          // If failed at last resort (EC2 Metadata check) expose a meaningful error
+          // with link to AWS documentation
+          // Otherwise, it's likely that user relied on some AWS creds, which appeared not correct
+          // therefore expose an AWS message directly
+          let bottomError = err;
+          while (bottomError.originalError && !bottomError.message.startsWith('EC2 Metadata')) {
+            bottomError = bottomError.originalError;
+          }
+
+          const errorMessage = bottomError.message.startsWith('EC2 Metadata')
+            ? [
+                'AWS provider credentials not found.',
+                ' Learn how to set up AWS provider credentials',
+                ` in our docs here: <${chalk.green('http://slss.io/aws-creds-setup')}>.`,
+              ].join('')
+            : bottomError.message;
+          message = errorMessage;
+          // We do not want to trigger the retry mechanism for credential errors
+          return BbPromise.reject(
+            Object.assign(new ServerlessError(errorMessage, err.code), {
+              providerError: Object.assign({}, err, { retryable: false }),
+            })
+          );
+        }
+
+        return BbPromise.reject(
+          Object.assign(new ServerlessError(message, err.code), {
+            providerError: err,
+          })
+        );
+      });
+    }).then((data) => {
+      const result = BbPromise.resolve(data);
+      return result;
+    })
+  );
+
+  return request;
+}
+
+function canUseS3TransferAcceleration(service, method) {
+  // TODO enable more S3 APIs?
+  return service === 'S3' && ['upload', 'putObject'].indexOf(method) !== -1;
+}
+
+function enableS3TransferAcceleration(credentials) {
+  log('Using S3 Transfer Acceleration Endpoint...');
+  credentials.useAccelerateEndpoint = true; // eslint-disable-line no-param-reassign
+}
+
+awsRequest.memoized = memoize(awsRequest);
```

Comment:
There are few issues with this memoization setup:
1.  By default `memoizee` will memoize arguments by _instance_ and not by content.
2. `memoizee` by default memoizes only against required function arguments. 
3. Async functions need to be communicated with `promise: true` option

So configuration as put cannot work well.

To fix it, the only way is to provide `normalizer` option (which technically is about resolving a cache key) as below:

```javascript
const deepSortObjectByKey = require('../utils/deepSortObjectByKey');

awsRequest.memoized = memoize(awsRequest, { promise: true,  normalizer: ([serviceData, method, params = {}]) => {
  if (typeof serviceData === 'string') serviceData = { name: serviceData });
  return [deepSortObjectByKey(serviceData), method, deepSortObjectByKey(params)].join("|");
}) });
```

Reply:
Ok, I need to have a deeper look into it :smile: I definitely missed a couple things there.

Reply:
I changed it to match yours but using the new signature. I think the behavior should be ok now.